nmosnode/authclient.py

```diff
@@ -0,0 +1,193 @@
+        self.allowed_response = allowed_response
+        self.auth_method = auth_method
+
+        self.client_id = None
+        self.client_secret = None
+        self.bridge = IppmDNSBridge()
+        self._client_registry = {}
+        self.initialised = self.initialise()
+
+    def initialise(self):
+        """Check if credentials file already exists, meaning the device is already registered.
+        If not, register with Auth Server and write client credentials to file."""
+        try:
+            if os.path.isfile(CREDENTIALS_PATH):
+                logger.writeWarning("Credentials file already exists. Using existing credentials.")
+                self.client_id, self.client_secret = get_credentials_from_file(CREDENTIALS_PATH)
+            else:
+                logger.writeInfo("Registering with Authorization Server...")
+                reg_resp_json = self.send_oauth_registration_request()
+                self.write_credentials_to_file(reg_resp_json, CREDENTIALS_PATH)
+            return True
+        except Exception as e:
+            logger.writeError(
+                "Unable to initialise OAuth Client with client credentials. {}".format(e)
+            )
+            return False
+
+    def write_credentials_to_file(self, data, file_path):
+        try:
+            self.client_id = data.get('client_id')
+            self.client_secret = data.get('client_secret')
+            credentials = {
+                "client_id": self.client_id,
+                "client_secret": self.client_secret
+            }
+            with open(file_path, 'w') as f:
+                json.dump(credentials, f)
+            os.chmod(file_path, 0o600)
+            return True
+        except OSError as e:
+            logger.writeError(
+                "Could not write OAuth client credentials to file {}. {}".format(file_path, e)
+            )
+            raise
+
+    def send_oauth_registration_request(self):
+        try:
+            href = self.bridge.getHref(MDNS_SERVICE_TYPE)
+            registration_href = urljoin(href, REGISTRATION_ENDPOINT)
+            logger.writeDebug('Registration endpoint href is: {}'.format(registration_href))
+
+            data = {
+                "client_name": self.client_name,
+                "client_uri": self.client_uri,
+                "scope": self.allowed_scope,
+                "redirect_uri": self.redirect_uri,
+                "grant_type": self.allowed_grant,
+                "response_type": self.allowed_response,
+                "token_endpoint_auth_method": self.auth_method
+            }
+
+            # Decide how Basic Auth details are retrieved - user input? Retrieved from file?
+            reg_resp = requests.post(
+                registration_href,
+                data=data,
+                auth=('dannym', 'password'),
+                timeout=0.5,
+                proxies={'http': ''}
+            )
+            reg_resp.raise_for_status()  # Raise error if status !=201
```

Comment:
If the Node had previously registered but has since restarted, would it not get a 200?

Reply:
This portion of the code shouldn't run if the node is already registered (i.e. the presence of a credentials file)